### PR TITLE
Fix source version check

### PR DESCRIPTION
**Pull request by msaroufim**
@crcrpar can you try this instead?

Fixes https://github.com/pytorch/ao/pull/679#issuecomment-2290444713

This just basically takes advantage of the fact that every PyTorch version has the format `X.Y.Z` and then some stuff

```python

# Test cases
import torch
from torchao.utils import torch_version_at_least

torch.__version__ = "2.5.0a0+git9f17037"
print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True
print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True


torch.__version__ = "2.5.0.dev20240708+cu121"
print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True
print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True


torch.__version__ = "2.5.0"
print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True

torch.__version__ = "2.4.0"
print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be False

```

### CODE REVIEW

**Comment by jerryzh168** [comment]
> @crcrpar can you try this instead?
> 
> Fixes [#679 (comment)](https://github.com/pytorch/ao/pull/679#issuecomment-2290444713)
> 
> This just basically takes advantage of the fact that every PyTorch version has the format `X.Y.Z` and then some stuff
> 
> ```python
> # Test cases
> import torch
> from torchao.utils import torch_version_at_least
> 
> torch.__version__ = "2.5.0a0+git9f17037"
> print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True
> print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
> 
> 
> torch.__version__ = "2.5.0.dev20240708+cu121"
> print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True
> print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
> 
> 
> torch.__version__ = "2.5.0"
> print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
> print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be True
> 
> torch.__version__ = "2.4.0"
> print(f"{torch.__version__}: {torch_version_at_least('2.4.0')}")  # Should be True
> print(f"{torch.__version__}: {torch_version_at_least('2.5.0')}")  # Should be False
> ```

maybe add these examples to tests?

**Review by jerryzh168 — APPROVED** [review]

File: test/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import unittest
+from unittest.mock import patch
+from torchao.utils import torch_version_at_least
+
+class TestTorchVersionAtLeast(unittest.TestCase):
+    def test_torch_version_at_least(self):
+        test_cases = [
+            ("2.5.0a0+git9f17037", "2.5.0", True),
```
Comment (crcrpar) [inline test/test_utils.py:8]: I'm confused, as I thought this should be False instead.
shouldn't 2.5.0a0 and 2.5.0.dev be <2.5.0 but >2.4?
I might misunderstand `torch_version_at_least` usage though

Comment (msaroufim) [reply test/test_utils.py:8]: Strictly speaking this is an internal only utility that we leverage in case torch makes certain breaking changes or adds new features upstream. So I can't think of many cases where we'd do one thing with 2.5 nightlies vs 2.5 official releases.

Comment (crcrpar) [reply test/test_utils.py:8]: understood.

**Comment by felipemello1** [comment]
@msaroufim , when do you guys plan to make a new release? Is it monthly?

**Comment by msaroufim** [comment]
We do monthly, next one is scheduled around Sep 5
